### PR TITLE
feat: indexing of Celo transactions

### PR DESCRIPTION
**Pull request by fedor-ivn**
## Motivation

This PR adds support for custom fields from Celo blockchain and implements a
custom way to index Celo transactions.

#### Environment Variables

**New value** for `CHAIN_TYPE` variable: `celo`

**New variable**: `CELO_CORE_CONTRACTS`

The CELO network includes several core contracts that are central to its operation. These core contracts manage various aspects of the Celo network, including its stablecoins, governance, identity, and more.

All these contracts can be obtained by querying the `Registry` contract, which is deployed at the genesis block with the address `0x000000000000000000000000000000000000ce10`. This contract serves as a repository for the addresses of all core contracts on the Celo network.

While it's possible to process events from the Registry contract to keep information on core contracts up to date, implementing this in the current Blockscout indexer is not trivial. Fortunately, core contracts are not updated frequently. Thus, once fetched, the addresses typically remain unchanged. To simplify, I decided to store information about core contracts under the `CELO_CORE_CONTRACTS` variable, which is a JSON object of a specified structure.

In case the contracts are updated, maintaining the Blockscout instance will require updating the `CELO_CORE_CONTRACTS` variable and manually setting blocks for refetch (all blocks produced after the core contracts' update).

The structure of the JSON value for the `CELO_CORE_CONTRACTS` variable is detailed in this [gist](https://gist.github.com/fedor-ivn/c77030d09bd05349710450e0fcf00a1f). 

The JSON structure in the gist contains two main sections: `addresses` and `events`.

- **addresses**: This section maps contract names to the list of respective addresses and the block number at which they were updated. For example:
  - `"Accounts"`: `[{ "address": "0xed7f51a34b4e71fbe69b3091fcf879cd14bd73a9", "updated_at_block_number": 574 }]`
  - `"Election"`: `[{ "address": "0x1c3edf937cfc2f6f51784d20deb1af1f9a8655fa", "updated_at_block_number": 592 }]`
- **events**: This section maps contract addresses to the events associated with them, including any parameters and the block numbers at which these events were emitted. For example:
  - `"EpochRewards"`: `{ "0xb10ee11244526b94879e1956745ba2e35ae2ba20": { "CarbonOffsettingFundSet": [{ "address": "0x22579ca45ee22e2e16ddf72d955d6cf4c767b0ef", "updated_at_block_number": 15049265 }] } }`
  - `"FeeHandler"`: `{ "0x90d94229623a0a38826a4a7557a6d79acde43f76": { "BurnFractionSet": [{ "updated_at_block_number": 19732579, "value": 0.7999999999999999 }] } }`

In the [next PR](https://github.com/blockscout/blockscout/pull/9944), there is already an implemented [mix task](https://github.com/blockscout/blockscout/blob/162756336e448bf5638858856accf36167e8656b/apps/explorer/lib/fetch_celo_core_contracts.ex) that automatically assembles this variable by indexing the entire chain from the `0` to the `latest` block number.

These changes are also reflected in docs, see https://github.com/blockscout/docs/pull/276.

#### Transactions

New fields specific to the CELO network are now supported. Here is an outline of how the fields from the RPC node are mapped to the fields in the `transactions` table:

- `feeCurrency` → `gas_token_contract_address_hash`
- `gatewayFeeRecipient` → `gas_fee_recipient_address_hash` (this field is scheduled for deprecation as stated in the [CELO docs](https://docs.celo.org/protocol/transaction/transaction-types#-legacy-transaction-0-1))
- `gatewayFee` → `gateway_fee` (this field is scheduled for deprecation as stated in the [CELO docs](https://docs.celo.org/protocol/transaction/transaction-types#-legacy-transaction-0-1))

The token used for gas payment (`feeCurrency`) is indexed within the block fetcher process.

#### Token-Duality Feature

Native chain asset transfers are treated as CELO ERC-20 token transfers. These token transfers are "artificial," meaning there is no actual `Transfer` event emitted by the CELO ERC-20 contract that corresponds to the token transfer.

**⚠️ Important Note:** Since there is no actual log associated with such token transfers, these transfers are saved with a negative log index in the `token_transfers` table.

Indexing of native chain asset transfers as CELO ERC-20 token transfers is handled by the block fetcher and internal transactions fetcher.

P.S: [Example of transaction][token-duality-tx] containing token transfers derived from native CELO transfer. Notice that the logs tab is empty.

[token-duality-tx]: https://celo-alfajores.blockscout.com/tx/0x424371ea01885b25e05cb9c8d48abf4a48364e45dbf49e2529d32c11e9d9494d?tab=index

#### API Updates

The following endpoints have been extended to include a `celo` element in the response, offering additional information about the token used for paying gas:
- `/api/v2/transactions`
- `/api/v2/transactions/{transaction_hash}`
- `/api/v2/addresses/{address_hash}/transactions`
- `/api/v2/main-page/transactions`

Example of the `celo` section in the response:
```json
"celo": {
  "gas_token": {
    "address": "0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1",
    "circulating_market_cap": null,
    "decimals": "18",
    "exchange_rate": null,
    "holders": "3778",
    "icon_url": null,
    "name": "Celo Dollar",
    "symbol": "cUSD",
    "total_supply": "7090075332805485657875961",
    "type": "ERC-20",
    "volume_24h": null
  }
}
```

P.S: [Example of transaction][gas-token-tx] utilizing gas token on Alfajores test net.

[gas-token-tx]: https://celo-alfajores.blockscout.com/api/v2/transactions/0xc27c7037169347226b2cbdef1379dd34e36c0669d13349cf8131441321578dc9

## Checklist for your Pull Request (PR)

  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [x] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [x] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/indexer/test/indexer/block/fetcher_test.exs [hunk 1]
```diff
@@ -279,6 +280,48 @@ defmodule Indexer.Block.FetcherTest do
...
+            transaction =
+              Map.merge(
+                base_transaction,
+                if(
+                  chain_type == "celo",
+                  do: celo_fields_transaction,
+                  else: %{}
+                )
+              )
+
```
Comment (fedor-ivn) [inline apps/indexer/test/indexer/block/fetcher_test.exs:324]: Is it okay to integrate check of celo transaction fields into the existing test?

File: apps/indexer/test/indexer/block/fetcher_test.exs [hunk 2]
```diff
@@ -614,6 +607,39 @@ defmodule Indexer.Block.FetcherTest do
           assert fifth_address.fetched_coin_balance_block_number == block_number
 
         EthereumJSONRPC.Nethermind ->
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/indexer/test/indexer/block/fetcher_test.exs:609]: Because of this, I had nothing to do, but put my celo-changes to `EthereumJSONRPC.Nethermind` branch

File: apps/indexer/test/indexer/block/fetcher_test.exs (hunk 1 above)
Comment (nikitosing) [reply apps/indexer/test/indexer/block/fetcher_test.exs:324]: I thinks it's ok

File: apps/explorer/lib/explorer/chain/cache/celo_core_contracts.ex
Comment (fedor-ivn) [inline apps/explorer/lib/explorer/chain/cache/celo_core_contracts.ex:1]: I think hard-coded values are fine at least for v1, because they are really not a subject to change. 

For instance, on [Celo mainnet](https://explorer.celo.org/mainnet/address/0x000000000000000000000000000000000000ce10/transactions#address-tabs), the last transaction to Registry contract was done 4 years ago. On alfajores, Registry contract was not even updated -- e.g it returns nil-address (0x000...) for the core contract names.

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex [hunk 3]
```diff
@@ -406,4 +434,40 @@ defmodule Indexer.Fetcher.InternalTransaction do
...
+        else
+          error -> Logger.error("Failed to cast string to hash: #{inspect(error)}")
+        end
+      end)
+      |> MapSet.new()
+    end
+
+    defp async_import_celo_token_balances(%{token_transfers: token_transfers, tokens: tokens}) do
+      case tokens do
+        [token] ->
```
Comment (nikitosing) [inline apps/indexer/lib/indexer/fetcher/internal_transaction.ex:461]: Why not to case `token_transfers`?

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex [hunk 4]
```diff
@@ -406,4 +434,40 @@ defmodule Indexer.Fetcher.InternalTransaction do
...
+            token_id: nil
+          }
+        else
+          error -> Logger.error("Failed to cast string to hash: #{inspect(error)}")
+        end
+      end)
+      |> MapSet.new()
+    end
+
+    defp async_import_celo_token_balances(%{token_transfers: token_transfers, tokens: tokens}) do
```
Comment (nikitosing) [inline apps/indexer/lib/indexer/fetcher/internal_transaction.ex:459]: I would suggest to define functions in order in which they are called. I.e. firstly define `async_import_celo_token_balances` and then `to_address_token_balances`

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex (hunk 3 above)
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/fetcher/internal_transaction.ex]: Yeah, make sense. Please, check https://github.com/blockscout/blockscout/pull/9713/commits/222febab551e9ee6b9fe6901e2916970706e2a05 to verify if I implemented your suggestions correctly

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex (hunk 4 above)
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/fetcher/internal_transaction.ex]: Also fixed in https://github.com/blockscout/blockscout/pull/9713/commits/222febab551e9ee6b9fe6901e2916970706e2a05

File: .github/workflows/config.yml [hunk 5]
```diff
@@ -51,7 +51,7 @@ jobs:
 
             // Add/remove CI matrix chain types here
             const defaultChainTypes = ["default"];
-            const chainTypes = ["ethereum", "polygon_zkevm", "rsk", "stability", "filecoin", "optimism", "arbitrum"];
+            const chainTypes = ["ethereum", "polygon_zkevm", "rsk", "stability", "filecoin", "optimism", "arbitrum", "celo"];
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply .github/workflows/config.yml:54]: Sure! Done in the PR https://github.com/blockscout/docs/pull/276

File: apps/indexer/lib/indexer/transform/celo/transaction_gas_tokens.ex [hunk 6]
```diff
@@ -0,0 +1,29 @@
+defmodule Indexer.Transform.Celo.TransactionGasTokens do
+  @moduledoc """
+  Helper functions for extracting tokens specified as gas fee currency.
+  """
+
+  @doc """
+  Parses transactions and extracts tokens specified as gas fee currency.
+  """
+  def parse(transactions) do
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/transform/celo/transaction_gas_tokens.ex:9]: Do you agree with such spec?

```elixir
 @spec parse([
          %{
            optional(:gas_token_contract_address_hash) => Hash.Address.t() | nil
          }
        ]) :: [
          %{
            contract_address_hash: String.t(),
            type: String.t()
          }
        ]
  def parse(transactions) do
```

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 7]
```diff
@@ -150,6 +153,20 @@ defmodule Indexer.Block.Fetcher do
...
+             Application.get_env(:explorer, :chain_type) == :celo,
+             do: CeloTransactionTokenTransfers.parse_transactions(transactions_with_receipts),
+             else: %{token_transfers: [], tokens: []}
+           ),
+         celo_gas_tokens =
+           if(
+             Application.get_env(:explorer, :chain_type) == :celo,
+             do: CeloTransactionGasTokens.parse(transactions_with_receipts),
+             else: []
+           ),
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/block/fetcher.ex:167]: Ok, done it. Btw, I assumed transformers do not produce side effects

File: apps/explorer/lib/explorer/chain/import/runner/transactions.ex [hunk 8]
```diff
@@ -107,6 +107,7 @@ defmodule Explorer.Chain.Import.Runner.Transactions do
     )
   end
 
+  # todo: avoid code duplication
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/import/runner/transactions.ex:110]: Do you have any ideas on how I can refactor that?

Should we create a new refactoring issue or target this in the current PR? What do you think?

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex [hunk 9]
```diff
@@ -290,6 +317,8 @@ defmodule Indexer.Fetcher.InternalTransaction do
           address_hash_to_fetched_balance_block_number: address_hash_to_block_number
         })
 
+        async_import_celo_token_balances(celo_token_transfers_params)
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/fetcher/internal_transaction.ex:320]: > The definition of this function is under compile env.

Indeed. However, there is a dummy function defined in else branch (for non-`:celo` chain type). Or you mean it's better to put a call to `async_import_celo_token_balances/1` under an if condition and remove conditional definition of this function?

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 10]
```diff
@@ -192,7 +209,8 @@ defmodule Indexer.Block.Fetcher do
              blocks_params: blocks,
              logs_params: logs,
              transactions_params: transactions_with_receipts,
-             withdrawals: withdrawals_params
+             withdrawals: withdrawals_params,
+             celo_token_transfers: token_transfers
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/block/fetcher.ex:213]: Right, I see... Thanks!

[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/block/fetcher.ex:213]: Btw, do we need to filter *real* celo token transfers and update balances for them? I mean, you are right, we derive celo token transfers from transactions, but we don't do it vice versa

File: apps/explorer/lib/explorer/chain/import/runner/transactions.ex (hunk 8 above)
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/explorer/lib/explorer/chain/import/runner/transactions.ex:110]: I don't think, we should apply this in the current PR. But makes sense, creating refactoring issue with minor priority to the backlog.

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex (hunk 9 above)
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/indexer/lib/indexer/fetcher/internal_transaction.ex:317]: Yes, I mean shouldn't we wrap its call to `if statement`. Seems, this function to be called for any chain type (or I missing smthng), no? However, it is defined only for compile-time `:celo` chain type, which is correct.

[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/fetcher/internal_transaction.ex:317]: I suppose no, since we have a so-called "default branch" down below. Thus, under non-`:celo` chain types the call to this function will be simply ignored. If this behavior seems to be implicit, I can implement your suggestion with `if` statement.  

https://github.com/blockscout/blockscout/blob/a2178485763b44a88f7a13b8c219c7cb48cf354c/apps/indexer/lib/indexer/fetcher/internal_transaction.ex#L467

File: apps/indexer/test/indexer/block/realtime/fetcher_test.exs [hunk 11]
```diff
@@ -63,6 +71,13 @@ defmodule Indexer.Block.Realtime.FetcherTest do
 
       ReplacedTransaction.Supervisor.Case.start_supervised!()
 
+      # In CELO network, there is a token duality feature where CELO can be used
+      # as both a native chain currency and as an ERC-20 token. Transactions
+      # that transfer CELO are also counted as token transfers, and the
+      # TokenInstance fetcher is called. However, for simplicity, we disable it
+      # in this test.
+      Application.put_env(:indexer, Indexer.Fetcher.TokenInstance.Realtime.Supervisor, disabled?: true)
```
Comment (fedor-ivn) [inline apps/indexer/test/indexer/block/realtime/fetcher_test.exs:79]: What do you think about this workaround?

Comment (fedor-ivn) [reply apps/indexer/test/indexer/block/realtime/fetcher_test.exs:79]: @vbaranov @nikitosing

File: apps/explorer/lib/explorer/chain/cache/celo_core_contracts.ex
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/cache/celo_core_contracts.ex:1]: The workflow regarding core contracts was updated. Please, read the new PR description

File: apps/indexer/lib/indexer/transform/token_transfers.ex [hunk 12]
```diff
@@ -60,14 +60,7 @@ defmodule Indexer.Transform.TokenTransfers do
...
-    |> Enum.filter(fn token_transfer ->
-      token_transfer.to_address_hash == burn_address_hash_string() ||
-        token_transfer.from_address_hash == burn_address_hash_string()
-    end)
-    |> Enum.map(fn token_transfer ->
-      token_transfer.token_contract_address_hash
-    end)
-    |> Enum.uniq()
+    |> filter_tokens_for_supply_update()
     |> TokenTotalSupplyUpdater.add_tokens()
```
Comment (nikitosing) [inline apps/indexer/lib/indexer/transform/token_transfers.ex:64]: What is the reason for this change?

File: apps/indexer/lib/indexer/fetcher/token_total_supply_updater.ex [hunk 13]
```diff
@@ -5,6 +5,8 @@ defmodule Indexer.Fetcher.TokenTotalSupplyUpdater do
 
   use GenServer
 
+  import Explorer.Chain.SmartContract, only: [burn_address_hash_string: 0]
```
Comment (nikitosing) [inline apps/indexer/lib/indexer/fetcher/token_total_supply_updater.ex:8]: Seems that this import is unused

File: apps/indexer/lib/indexer/transform/token_transfers.ex (hunk 12 above)
Comment (nikitosing) [reply apps/indexer/lib/indexer/transform/token_transfers.ex:64]: Understood

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex [hunk 14]
```diff
@@ -10,15 +10,19 @@ defmodule Indexer.Fetcher.InternalTransaction do
...
 
   alias EthereumJSONRPC.Utility.RangesHelper
   alias Explorer.Chain
-  alias Explorer.Chain.Block
+  alias Explorer.Chain.{Block, Hash}
   alias Explorer.Chain.Cache.{Accounts, Blocks}
   alias Indexer.{BufferedTask, Tracer}
   alias Indexer.Fetcher.InternalTransaction.Supervisor, as: InternalTransactionSupervisor
-  alias Indexer.Transform.Addresses
+  alias Indexer.Transform.{Addresses, AddressTokenBalances}
```
Comment (nikitosing) [inline apps/indexer/lib/indexer/fetcher/internal_transaction.ex:24]: Let's move `AddressTokenBalances` alias under compile env on 442 line

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 15]
```diff
@@ -151,6 +154,11 @@ defmodule Indexer.Block.Fetcher do
          %{logs: logs, receipts: receipts} = receipt_params,
          transactions_with_receipts = Receipts.put(transactions_params_without_receipts, receipts),
          %{token_transfers: token_transfers, tokens: tokens} = TokenTransfers.parse(logs),
+         %{token_transfers: celo_native_token_transfers, tokens: celo_tokens} =
+           CeloTransactionTokenTransfers.parse_transactions(transactions_with_receipts),
+         celo_gas_tokens = CeloTransactionGasTokens.parse(transactions_with_receipts),
```
Comment (nikitosing) [inline apps/indexer/lib/indexer/block/fetcher.ex:159]: Shouldn't we send balances (sender and miner) related to gas token to update? or gas token transfers tracked in logs?

File: apps/indexer/lib/indexer/fetcher/token_total_supply_updater.ex (hunk 13 above)
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/fetcher/token_total_supply_updater.ex:8]: good catch :+1:

File: apps/indexer/lib/indexer/fetcher/internal_transaction.ex (hunk 14 above)
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/fetcher/internal_transaction.ex:24]: :+1:

File: apps/indexer/lib/indexer/block/fetcher.ex (hunk 15 above)
Comment (fedor-ivn) [reply apps/indexer/lib/indexer/block/fetcher.ex:159]: Good question. If I understand correctly, when a token is used for paying gas, it generates the corresponding Transfer logs. However, I will clarify this with the Celo team to be sure.

Comment (fedor-ivn) [reply apps/indexer/lib/indexer/block/fetcher.ex:159]: Celo team confirmed that gas token transfers appear in the logs. Thus, I suppose we don't need to implement any custom logic for balance update here.

File: apps/explorer/test/explorer/chain/cache/celo_core_contracts_test.ex [hunk 16]
```diff
@@ -0,0 +1,75 @@
+defmodule Explorer.Chain.Cache.CeloCoreContractsTest do
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/test/explorer/chain/cache/celo_core_contracts_test.ex]: Processed in edbc604

File: apps/explorer/lib/explorer/chain/cache/celo_core_contracts.ex [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+    },
+    epoch_rewards: %{
+      carbon_offsetting_fund_set: "CarbonOffsettingFundSet"
+    }
+  }
+
+  def atom_to_contract_name, do: @atom_to_contract_name
+  def atom_to_contract_event_names, do: @atom_to_contract_event_names
+
+  def get_event(contract_atom, event_atom, block_number) do
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/cache/celo_core_contracts.ex:84]: Processed in edbc604

**Review by nikitosing — APPROVED** [review]
